### PR TITLE
Add Support for TMS 2.0

### PR DESCRIPTION
**Pull request by dchirst**
# Add TMS 2.0 Support

Specification: http://www.opengis.net/doc/IS/tms/2.0

I am following the change log here: https://docs.ogc.org/is/17-083r4/21-066r1.html#Clause_Critical

High level changes: 
- [Addition of a `uri` parameter for TileMatrixSets registered on the official OGC NA TileMatrixSets registry](https://docs.ogc.org/is/17-083r4/21-066r1.html#_identifying_with_uri_a_well_known_tilematrixset_in_an_authoritative_registry)
- [`identifier` renamed to `id`](https://docs.ogc.org/is/17-083r4/21-066r1.html#_json_encoding_rules_to_derive_a_more_natural_json_encoding_from_uml)
- [`supportedCRS` renamed to `crs`](https://docs.ogc.org/is/17-083r4/21-066r1.html#_renaming_supportedcrs_to_crs)
- [`tileMatrix` renamed to `tileMatrices`](https://docs.ogc.org/is/17-083r4/21-066r1.html#_json_encoding_rules_to_derive_a_more_natural_json_encoding_from_uml)
- [`topLeftCorner` renamed to `pointOfOrigin`](https://docs.ogc.org/is/17-083r4/21-066r1.html#_replacing_topleftcorner_by_pointoforigin)
- [Add optional parameters `cellSize` and `cornerOfOrigin` to TileMatrix](https://docs.ogc.org/is/17-083r4/21-066r1.html#_adding_cellsize_and_corneroforigin)
- [Remove `type` object properties](https://docs.ogc.org/is/17-083r4/21-066r1.html#_removing_type_object_properties)
- [Added optional axis ordering](https://docs.ogc.org/is/17-083r4/21-066r1.html#_adding_optional_orderedaxes_to_highlight_crs_axis_ordering
)

Changes that are ignored:
1. [`variableMatrixWidth` renamed to `variableMatrixWidths`](https://docs.ogc.org/is/17-083r4/21-066r1.html#_json_encoding_rules_to_derive_a_more_natural_json_encoding_from_uml) - There are no references to `variableMatrixWidth` except in a comment that tells pydantic to explicitly forbid it. I assume that this parameter was omitted for a reason
2. [Rename `TileMatrixSetLink` to `TileSetMetadata`](https://docs.ogc.org/is/17-083r4/21-066r1.html#_data_model_for_tileset_metadata) - likewise, no reference to `TileMatrixSetLink` exist in morecantile
3. [OWS Common Dependency Removed](https://docs.ogc.org/is/17-083r4/21-066r1.html#_ows_common_dependency_removed) - No reference to OWS common in morecantile so there is no dependency already!
4. [Make `boundingBox` optional](https://docs.ogc.org/is/17-083r4/21-066r1.html#_making_boundingbox_optional) - bounding box is already optional in morecantile!
5. [Removing the JSON-LD](https://docs.ogc.org/is/17-083r4/21-066r1.html#_removing_the_json_ld) - morecantile doesn't support JSON-LD anyway
6. [6.10, 6.11, 6.12](https://docs.ogc.org/is/17-083r4/21-066r1.html#_new_annex_with_variable_width_tilematrixsets_definitions) - this is just admin, no real changes to morecantile
7. [TileSetLink -> TileMetadata](https://docs.ogc.org/is/17-083r4/21-066r1.html#_data_model_for_tileset_metadata) - there doesn't seem to be any reference to TileSetLink in the current version of morecantile, so I'm assuming we don't need to implement this
8. [Align with Abstract](https://docs.ogc.org/is/17-083r4/21-066r1.html#_align_with_the_abstract_specification_topic_22) - I _think_ this is administrative, with no change to the code
9. [JSON encoding for TileSet metadata](https://docs.ogc.org/is/17-083r4/21-066r1.html#_xml_and_json_encoding_for_tileset_metadata) - Again, there doesn't seem to be much reference to Tileset in morecantile, so I'm assuming this isn't necessary. However, I'm not sure about this and so should probably get a second opinion

- [Support direct CRS definition](https://docs.ogc.org/is/17-083r4/21-066r1.html#_allowing_embedded_wkt2_crs_definition_in_addition_to_crs_by_reference) - Pyproj supports WKT2 just fine. However, I haven't found a way to implement ISO 19115 MD_ReferenceSystem

This is currently a draft to address to #91 

TODO: 
- Docs

### CODE REVIEW

**Comment by vincentsarago** [comment]
🥳 thanks for starting this 🙏

**Comment by dchirst** [comment]
@vincentsarago hi 👋 ! I was wondering if I could ask your advice on a couple of these changes:

1. There is now a TileSetMetadata definition, replacing the TileMatrixSetLink. Do we want to implement this in morecantile or should we leave it without?
2. All the tests are passing, except for the tests in `tests/test_mercantile_conform.py`. These tests fail due to a rounding error, where the values differ by around 1e-7. Does this matter? If so, do you have any idea what could be causing these rounding errors?
3. TMS 2.0 now supports direct CRS definitions, rather than CRS by reference. This includes ISO 19115 MD_ReferenceSystem and WKT2. Pyproj supports user input of WKT2 but not MD_ReferenceSystem? Do you know anywhere I could find an ISO 19115 parser in Python?

Thanks!

**Comment by vincentsarago** [comment]
> There is now a TileSetMetadata definition, replacing the TileMatrixSetLink. Do we want to implement this in morecantile or should we leave it without?

I'm not sure to understand what's the `TileSetMetadata` in the OGC docs 🤷 . Is it a representation of the TMS document? 

> All the tests are passing, except for the tests in tests/test_mercantile_conform.py. These tests fail due to a rounding error, where the values differ by around 1e-7 (where the units are meters). Does this matter? If so, do you have any idea what could be causing these rounding errors?

We can round the number or use numpy almost equal https://github.com/developmentseed/morecantile/blob/644d3b70547420222d3c34071f2eb78da0e37646/tests/test_models.py#L189


> TMS 2.0 now supports direct CRS definitions, rather than CRS by reference. This includes ISO 19115 MD_ReferenceSystem and WKT2. Pyproj supports user input of WKT2 but not MD_ReferenceSystem? Do you know anywhere I could find an ISO 19115 parser in Python?

👍, no I don't and I think it's ok if we don't support all the things!

**Comment by dchirst** [comment]
> I'm not sure to understand what's the TileSetMetadata in the OGC docs 🤷 . Is it a representation of the TMS document?

I think TileSetMetadata is a standardised way to add auxiliary information, such as `version`, `keywords`, and `pointOfContact`.

> We can round the number or use numpy almost equal

The code already rounds the number to 7 decimal places, which is juuuuuuust precise enough that the tests fail. I just didn't want to lower the precision to 6 decimal places without good reason!

> 👍, no I don't and I think it's ok if we don't support all the things!

Fabulous, that makes it a lot easier! I can add a NotImplementedError just in case but I won't try and get it to work.

Thank you! It's getting close to being done - I just need to add a couple more tests and add info to the docs

**Comment by vincentsarago** [comment]
I pushed directly in you branch
- ref #106 (get tms bbox from the first matrix)
- update from `main`

File: morecantile/data1.0/CanadianNAD83_LCC.json [hunk 1]
```diff
@@ -0,0 +1,358 @@
+{
```
Comment (vincentsarago) [inline morecantile/data1.0/CanadianNAD83_LCC.json:1]: we don't really have to keep them! They can be found in https://github.com/vincentsarago/tilematrixsets

**Comment by vincentsarago** [comment]
@dchirst what do you think about adding a `from_v1` class method to create a TMS 2.0 from a v1 document? (we could do this in another PR)

**Comment by dchirst** [comment]
> @dchirst what do you think about adding a `from_v1` class method to create a TMS 2.0 from a v1 document? (we could do this in another PR)

@vincentsarago I think that could be a good idea. Happy to do it in this PR if you want! I think I'm just about done with the code, and have been drafting the updated docs. Would you like to review now and see if we can merge it in this week?

**Review by vincentsarago — APPROVED** [review]
all good 🙏 

Thanks for an incredible work @dchirst 

Next step is: Documentation 😄 but I think we can go ahead and merge this PR

**Comment by dchirst** [comment]
Amazing! My goal is to finish some documentation before the end of tomorrow. I haven't got the approval to merge the branch in myself, so feel free to merge it if you'd like!

**Comment by vincentsarago** [comment]
> Amazing! I haven't got the approval to merge the branch in myself, so feel free to merge it if you'd like! (remember to bump the version - i'm not sure if it's automatic!)

it's not automatic, I'm not planning to `release` this right now, we have others things to change too. Because this will be a new `major` version I want to make sure we do all the breaking change we need 😄 

> I actually just created a first draft version of the docs for TMS v2. Feel free to have a look and let me know if there's anything to add (I was surprised by how little there was to say)

That's perfect for now 🙏

File: morecantile/models.py [hunk 2]
```diff
@@ -277,13 +339,16 @@ def custom(
...
+            "crs": crs,
+            "tileMatrices": [],
             "_geographic_crs": geographic_crs,
         }
 
-        is_inverted = crs_axis_inverted(crs)
+        if ordered_axes:
+            is_inverted = ordered_axis_inverted(ordered_axes)
+        else:
+            is_inverted = crs_axis_inverted(crs)
```
Comment (vincentsarago) [inline morecantile/models.py:351]: @dchirst should we set `orderedAxes` in the TMS document based on the CRS information? 

To me it feels weird that one could pass a `ordered_axes` or ["Y", "X") with a CRS that would be `["X", "Y"]` ?

File: morecantile/data/WGS1984Quad.json [hunk 3]
```diff
@@ -1,255 +1,325 @@
...
-                -180
-            ],
-            "tileWidth": 256,
-            "tileHeight": 256,
-            "matrixWidth": 262144,
-            "matrixHeight": 131072
-        }
-    ]
-}
+  "id": "WorldCRS84Quad",
```
Comment (vincentsarago) [inline morecantile/data/WGS1984Quad.json:2]: @dchirst sorry to just catch this now, but any good reason for this to go from `WGS1984Quad` to `WorldCRS84Quad`? 

I believe the ID should be `WGS1984Quad`. The URI is fine being `http://www.opengis.net/def/tilematrixset/OGC/1.0/WorldCRS84Quad` as discussed in
ref: https://github.com/opengeospatial/2D-Tile-Matrix-Set/issues/60

**Pull request merged**